bson/src/main/scala/org/mongodb/scala/bson/codecs/IterableCodec.scala

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2016 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mongodb.scala.bson.codecs
+
+import java.util.UUID
+
+import scala.collection.mutable
+
+import org.bson._
+import org.bson.codecs._
+import org.bson.codecs.configuration.CodecRegistry
+
+/**
+ * IterableCodec companion object
+ *
+ * @since 1.2
+ */
+object IterableCodec {
+
+  def apply(registry: CodecRegistry, bsonTypeClassMap: BsonTypeClassMap): IterableCodec = apply(registry, bsonTypeClassMap, None)
+
+  def apply(registry: CodecRegistry, bsonTypeClassMap: BsonTypeClassMap, valueTransformer: Option[Transformer]): IterableCodec = {
+    new IterableCodec(registry, bsonTypeClassMap, valueTransformer.getOrElse(DEFAULT_TRANSFORMER))
+  }
+
+  private val DEFAULT_TRANSFORMER = new Transformer() {
+    def transform(objectToTransform: Object): Object = objectToTransform
+  }
+}
+
+/**
+ * Encodes and decodes `Iterable` objects.
+ *
+ * @since 1.2
+ */
+@SuppressWarnings(Array("rawtypes"))
+case class IterableCodec(registry: CodecRegistry, bsonTypeClassMap: BsonTypeClassMap, valueTransformer: Transformer) extends Codec[Iterable[_ <: Any]] {
+  lazy val bsonTypeCodecMap = new BsonTypeCodecMap(bsonTypeClassMap, registry)
+
+  override def decode(reader: BsonReader, decoderContext: DecoderContext): Iterable[_] =
+    readValue(reader, decoderContext).asInstanceOf[Iterable[_]]
+
+  override def encode(writer: BsonWriter, value: Iterable[_ <: Any], encoderContext: EncoderContext): Unit =
+    writeValue(writer, encoderContext, value)
+
+  override def getEncoderClass: Class[Iterable[_]] = classOf[Iterable[_]]
+
+  @SuppressWarnings(Array("unchecked", "rawtypes"))
+  private def writeValue[T](writer: BsonWriter, encoderContext: EncoderContext, value: T) {
+    value match {
+      case isNull if value == null => writer.writeNull() // scalastyle:ignore
+      case map: Map[_, _] =>
+        writeMap(writer, map.asInstanceOf[Map[String, Any]], encoderContext.getChildContext)
+      case list: Iterable[_] =>
+        writeIterable(writer, list, encoderContext.getChildContext)
+      case _ =>
+        val codec = registry.get(value.getClass).asInstanceOf[Encoder[T]]
+        encoderContext.encodeWithChildContext(codec, writer, value)
+    }
+  }
+
+  private def writeMap(writer: BsonWriter, map: Map[String, Any], encoderContext: EncoderContext) {
```

Comment:
Mainly to keep it DRY. Earlier I duplicated 90% of the code to handle maps in a MapCodec.  Here ordering becomes important - because a Map is an iterable, if the MapCodec is after the Iterable codec - this will infinite loop as this Iterable codec will try to handle it.